compiler/rustc_errors/src/annotate_snippet_emitter_writer.rs

```diff
@@ -329,20 +329,22 @@ impl AnnotateSnippetEmitter {
                     let substitutions = suggestion
                         .substitutions
                         .into_iter()
-                        .filter_map(|mut subst| {
+                        .filter(|subst| {
                             // Suggestions coming from macros can have malformed spans. This is a heavy
                             // handed approach to avoid ICEs by ignoring the suggestion outright.
                             let invalid =
                                 subst.parts.iter().any(|item| sm.is_valid_span(item.span).is_err());
                             if invalid {
                                 debug!("suggestion contains an invalid span: {:?}", subst);
                             }
-
+                            !invalid
+                        })
+                        .filter_map(|mut subst| {
                             // Assumption: all spans are in the same file, and all spans
                             // are disjoint. Sort in ascending order.
                             subst.parts.sort_by_key(|part| part.span.lo());
                             // Verify the assumption that all spans are disjoint
-                            assert_eq!(
+                            debug_assert_eq!(
```

Comment:
CC https://github.com/rust-lang/rust/pull/147849

Reply:
I have made local changes that work with #147849, so if it gets merged first, I can add those to this PR; if not, I can suggest those changes on the other PR.